### DISCUSSION TITLE
Secret API Keys in Plugin

### DISCUSSION THREAD

**Discussion by ReinhardtR** [runelite/runelite#15147] [Q&A] [answered]
Hello!

I'm not experienced in using REST API Keys in Java development, and  I don't know where to ask this. So here it goes.

Is it possible to use a secret key (like a REST API key) in my plugin, and keep that key secret? Or would it always be exposed in the build in some way?

**Comment by Enriath** [comment, marked as answer]
For hub plugins, the source must be made available (it's how they can be built and distributed in the first place), so yes, any key you put in there would be exposed for anyone to trivially abuse. (Even if the source was not available, it's pretty simple to pull strings from the compiled code).
If you want users to have a unique key to access a service (something a few hub plugins do already), it's better to ask for it as a config option that each user would have to enter.

**Reply by ReinhardtR** [reply]
That makes sense.

I want to make a API endpoint which receives a RSN, and will deny unauthenticated HTTP requests.

Is there a way to authenticate that a request is coming from a logged in RuneLite client?

**Reply by Enriath** [reply]
Given that you can't trust user input, no. A bad actor could easily fake the information if they chose to; that's not even a RuneLite thing, that's just a general software development thing.

**Reply by ReinhardtR** [reply]
Yeah, that was my worry.

Thanks for the support.
